core/client/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -40,20 +44,37 @@
 public final class AlluxioBlockStore {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
 
-  private static AlluxioBlockStore sClient = null;
+  // Only one block store will be kept for each master.
+  private static final Map<InetSocketAddress, AlluxioBlockStore> CACHED_BLOCK_STORES =
```

Comment:
Functionally it's the same either way, but it's cleaner and more testable if we store the block stores in the classes that need them. It also makes it more explicit what the object dependency graph looks like and makes it possible to manage block store lifecycles. Right now once a block store is created it will hang around forever even if nothing is using it any more. With block store users owning their own references to cached block stores, we could use reference counting to release unused block stores.